Insert panic when extractKeyExistsErrFromHandle

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```sql
CREATE TABLE `tt` (
  `c_int` int(11) DEFAULT NULL,
  `c_str` varbinary(40) NOT NULL,
  `c_timestamp` timestamp NULL DEFAULT NULL,
  `c_decimal` decimal(12,6) DEFAULT NULL,
  PRIMARY KEY (`c_str`(35)) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
insert into tt(c_int, c_str, c_timestamp, c_decimal) values (1, 'infallible elgamal', '2020-01-23 17:41:30', 6.423);
insert into tt (c_int, c_str, c_timestamp, c_decimal) values (35, 'infallible elgamal', '2020-02-25 03:28:53', 2.839);
```
### 2. What did you expect to see? (Required)
ERROR 1062 (23000): Duplicate entry 'infallible elgamal' for key 'tt.PRIMARY'
### 3. What did you see instead (Required)
ERROR 1105 (HY000): runtime error: slice bounds out of range [:35] with length 18
tidb panic
```
[2021/06/11 23:08:19.530 +08:00] [ERROR] [conn.go:801] ["connection running loop panic"] [conn=1317] [lastSQL="insert into tt (c_int, c_str, c_timestamp, c_decimal) values (35, 'infallible elgamal', '2020-02-25 03:28:53', 2.839)"] [err="runtime error: slice bounds out of range [:35] with length 18"] [stack="goroutine 22457793 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0x3be22f8, 0xc05e6a85a0, 0xc05e5ad500)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/server/conn.go:799 +0xf5
panic(0x3524a40, 0xc0334351b8)
\t/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/pingcap/tidb/store/driver/txn.extractKeyExistsErrFromHandle(0xc0329d21e0, 0x27, 0x30, 0xc02764e000, 0x35, 0x35, 0xc02777c300, 0x35, 0x0)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/driver/txn/error.go:98 +0xd89
github.com/pingcap/tidb/store/driver/txn.(*tikvTxn).extractKeyExistsErr(0xc0274eeab0, 0xc0329d21e0, 0x27, 0x30, 0x27026de, 0x365a8c0)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/driver/txn/txn_driver.go:227 +0x193
github.com/pingcap/tidb/store/driver/txn.(*tikvTxn).extractKeyErr(0xc0274eeab0, 0x3b97300, 0xc042beeb40, 0x3b97300, 0xc042beeb40)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/driver/txn/txn_driver.go:206 +0x7f
github.com/pingcap/tidb/store/driver/txn.(*tikvTxn).Commit(0xc0274eeab0, 0x3be22f8, 0xc0301d7500, 0x34fc640, 0x3bc1c00)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/driver/txn/txn_driver.go:67 +0x67
github.com/pingcap/tidb/session.(*LazyTxn).Commit(0xc05f6af410, 0x3be22f8, 0xc0301d7500, 0x0, 0x0)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/session/txn.go:354 +0x50b
github.com/pingcap/tidb/session.(*session).doCommit(0xc05f6af400, 0x3be22f8, 0xc0301d74a0, 0x0, 0x0)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/session/session.go:539 +0x6ef
github.com/pingcap/tidb/session.(*session).doCommitWithRetry(0xc05f6af400, 0x3be22f8, 0xc0301d74a0, 0x0, 0x0)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/session/session.go:603 +0x157
github.com/pingcap/tidb/session.(*session).CommitTxn(0xc05f6af400, 0x3be22f8, 0xc0301d6e10, 0x0, 0x0)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/session/session.go:665 +0x145
github.com/pingcap/tidb/session.autoCommitAfterStmt(0x3be22f8, 0xc0301d6e10, 0xc05f6af400, 0x0, 0x0, 0x3bf8238, 0xc042c129c0, 0x1, 0xc02714aad8)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/session/tidb.go:254 +0x42e
github.com/pingcap/tidb/session.finishStmt(0x3be22f8, 0xc0301d6e10, 0xc05f6af400, 0x0, 0x0, 0x3bf8238, 0xc042c129c0, 0x11990b8, 0x2)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/session/tidb.go:220 +0xcc
github.com/pingcap/tidb/session.runStmt(0x3be22f8, 0xc0301d6e10, 0xc05f6af400, 0x3bf8238, 0xc042c129c0, 0x0, 0x0, 0x0, 0x0)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/session/session.go:1664 +0x57a
github.com/pingcap/tidb/session.(*session).ExecuteStmt(0xc05f6af400, 0x3be22f8, 0xc0301d6e10, 0x3bfc3b8, 0xc027584b60, 0x0, 0x0, 0x0, 0x0)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/session/session.go:1548 +0xaab
github.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt(0xc05f709620, 0x3be22f8, 0xc0301d6e10, 0x3bfc3b8, 0xc027584b60, 0xc0265429e0, 0x3be22f8, 0xc0301d6e10, 0xf0fc)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/server/driver_tidb.go:218 +0x68
github.com/pingcap/tidb/server.(*clientConn).handleStmt(0xc05e5ad500, 0x3be2250, 0xc0301d6e10, 0x3bfc3b8, 0xc027584b60, 0x57ab7e0, 0x0, 0x0, 0x1, 0x0, ...)
\t/home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/server/conn.go:1691 +0x1bb
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc05e5ad500, 0x3be2250, 0xc044338040, 0xc03b3b0081, 0x75, 0x0, 0x0)
\t/home/jenkins/agent/workspace/optimization-"]

```
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
| Release Version: v5.1.0-20210611
Edition: Community
Git Commit Hash: 29bfbb28437609d0a524b7b84e6920dc765978c7
Git Branch: heads/refs/tags/v5.1.0-20210611
UTC Build Time: 2021-06-11 07:52:09
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |

--- a/executor/ddl_test.go
+++ b/executor/ddl_test.go
@@ -1526,14 +1526,3 @@ func (s *testRecoverTable) TestRenameMultiTables(c *C) {
 	tk.MustExec("drop database rename2")
 	tk.MustExec("drop database rename3")
 }
-
-// See https://github.com/pingcap/tidb/issues/24582
-func (s *testSuite6) TestDuplicatedEntryErr(c *C) {
-	tk := testkit.NewTestKit(c, s.store)
-	tk.MustExec("use test")
-	tk.MustExec("drop table if exists t1;")
-	tk.MustExec("create table t1(a int, b varchar(20), primary key(a,b(3)) clustered);")
-	tk.MustExec("insert into t1 values(1,'aaaaa');")
-	_, err := tk.Exec("insert into t1 values(1,'aaaaa');")
-	c.Assert(err.Error(), Equals, "[kv:1062]Duplicate entry '1-aaa' for key 'PRIMARY'")
-}

--- a/executor/insert_test.go
+++ b/executor/insert_test.go
@@ -1571,6 +1571,22 @@ func combination(items []string) func() []string {
 	}
 }
 
+// See https://github.com/pingcap/tidb/issues/24582
+func (s *testSuite10) TestDuplicatedEntryErr(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1;")
+	tk.MustExec("create table t1(a int, b varchar(20), primary key(a,b(3)) clustered);")
+	tk.MustExec("insert into t1 values(1,'aaaaa');")
+	err := tk.ExecToErr("insert into t1 values(1,'aaaaa');")
+	c.Assert(err.Error(), Equals, "[kv:1062]Duplicate entry '1-aaa' for key 'PRIMARY'")
+	err = tk.ExecToErr("insert into t1 select 1, 'aaa'")
+	c.Assert(err.Error(), Equals, "[kv:1062]Duplicate entry '1-aaa' for key 'PRIMARY'")
+	tk.MustExec("insert into t1 select 1, 'bb'")
+	err = tk.ExecToErr("insert into t1 select 1, 'bb'")
+	c.Assert(err.Error(), Equals, "[kv:1062]Duplicate entry '1-bb' for key 'PRIMARY'")
+}
+
 func (s *testSuite10) TestBinaryLiteralInsertToEnum(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
 	tk.MustExec(`use test`)

--- a/store/driver/txn/error.go
+++ b/store/driver/txn/error.go
@@ -94,7 +94,7 @@ func extractKeyExistsErrFromHandle(key kv.Key, value []byte, tblInfo *model.Tabl
 		if err != nil {
 			return genKeyExistsError(name, key.String(), err)
 		}
-		if col.Length > 0 {
+		if col.Length > 0 && len(str) > col.Length {
 			str = str[:col.Length]
 		}
 		valueStr = append(valueStr, str)

